Add --isolation rootless

nalind opened the pull request:
Add an isolation mode which also calls an OCI runtime, but with modified arguments, passing it a modified runtime configuration.  This mode does things which are currently not permitted for containers in SELinux policy.
We now reexec ourselves in a new user namespace — even when we're running with UID 0 — if we don't have `CAP_SYS_ADMIN`, so that we can use `unshare(2)` when setting up to run a runtime.

nalind committed:
Set BUILDAH_ISOLATION=rootless when running unprivileged

When we're run by an unprivileged user, default to BUILDAH_ISOLATION=rootless.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
.papr.sh: Also test with BUILDAH_ISOLATION=rootless

After the previously-defined tests are run, run the integration tests
again with rootless isolation.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
Add a value for IsolationOCIRootless

Add a new Isolation value: IsolationOCIRootless, for which we add a
global --rootless=true flag and a local --no-new-keyring flag when
creating a runtime container, and make some changes to the mounts list,
default namespacing configurations, and supplemental groups list.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
Skip certain tests when we're using "rootless" isolation

When we're testing with rootless isolation, skip tests that check that
we were able to influence namespace and control group settings, since we
hardwire namespace settings and don't try to configure control groups.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

rhatdan commented:
@nalind What AVCs are you seeing?

rhatdan commented:
DOes this replace https://github.com/projectatomic/buildah/pull/836 or in addition?
@giuseppe PTAL

nalind commented:
My preference would be to leave the SELinux policy alone, as that means we don't grant more privileges to containers, but these are logged in permissive mode:
```
----
time->Thu Jul 19 15:56:17 2018
type=AVC msg=audit(1532030177.684:21985): avc:  denied  { mount } for  pid=13547 comm="runc:[2:INIT]" name="/" dev="devpts" ino=1 scontext=system_u:system_r:container_t:s0:c508,c877 tcontext=system_u:object_r:devpts_t:s0 tclass=filesystem permissive=1
----
time->Thu Jul 19 15:56:17 2018
type=AVC msg=audit(1532030177.684:21984): avc:  denied  { mount } for  pid=13547 comm="runc:[2:INIT]" name="/" dev="tmpfs" ino=20748791 scontext=system_u:system_r:container_t:s0:c508,c877 tcontext=system_u:object_r:tmpfs_t:s0 tclass=filesystem permissive=1
----
time->Thu Jul 19 15:56:17 2018
type=AVC msg=audit(1532030177.684:21986): avc:  denied  { mounton } for  pid=13547 comm="runc:[2:INIT]" path="/tmp/buildah730234659/mnt/rootfs/dev/pts" dev="devpts" ino=1 scontext=system_u:system_r:container_t:s0:c508,c877 tcontext=system_u:object_r:devpts_t:s0 tclass=dir permissive=1
----
time->Thu Jul 19 15:56:17 2018
type=AVC msg=audit(1532030177.684:21987): avc:  denied  { mounton } for  pid=13547 comm="runc:[2:INIT]" path="/tmp/buildah730234659/mnt/rootfs/proc" dev="proc" ino=1 scontext=system_u:system_r:container_t:s0:c508,c877 tcontext=system_u:object_r:proc_t:s0 tclass=dir permissive=1
----
time->Thu Jul 19 15:56:17 2018
type=AVC msg=audit(1532030177.684:21988): avc:  denied  { remount } for  pid=13547 comm="runc:[2:INIT]" scontext=system_u:system_r:container_t:s0:c508,c877 tcontext=system_u:object_r:sysfs_t:s0 tclass=filesystem permissive=1
----
time->Thu Jul 19 15:56:17 2018
type=AVC msg=audit(1532030177.685:21989): avc:  denied  { open } for  pid=13547 comm="runc:[2:INIT]" path="/dev/pts/ptmx" dev="devpts" ino=2 scontext=system_u:system_r:container_t:s0:c508,c877 tcontext=system_u:object_r:devpts_t:s0 tclass=chr_file permissive=1
----
time->Thu Jul 19 15:56:17 2018
type=AVC msg=audit(1532030177.700:21990): avc:  denied  { setattr } for  pid=13547 comm="runc:[2:INIT]" name="0" dev="devpts" ino=3 scontext=system_u:system_r:container_t:s0:c508,c877 tcontext=system_u:object_r:devpts_t:s0 tclass=chr_file permissive=1
----
```

rhatdan commented:
If these are considered low enough risk to allow non priv users to mount, I don't see why SELinux should block.

allow container_t devpts_t:chr_file { open setattr };
allow container_t devpts_t:dir mounton;
allow container_t devpts_t:filesystem mount;
allow container_t proc_t:dir mounton;
allow container_t sysfs_t:filesystem remount;
allow container_t tmpfs_t:filesystem mount;
I guess the real question is why are these not labeled.

We have a devpts_t that is being mounted on?  Is this the systems /dev/pts?  
Mounting on /proc directories is probably not a concern?

[dwalsh@localhost psgo]$ sesearch -A -s container_t -p mount
allow container_domain container_file_t:filesystem { mount remount unmount };
allow container_t fusefs_t:filesystem mount; [ virt_sandbox_use_fusefs ]:True
allow svirt_sandbox_domain fusefs_t:filesystem mount; [ virt_sandbox_use_fusefs ]:True
allow svirt_sandbox_domain nfs_t:filesystem mount; [ virt_use_nfs ]:True

nalind commented:
These denials are seen when running `buildah run` inside of a container.  The /dev, /proc, and /sys mountpoints start off as ordinary directories in the rootfs that buildah has been building.  The devpts filesystem is being mounted with the "newinstance" flag.

rhatdan commented:
Is buildah passing in the mount label of the container it is running in?  I think the issue might be that we are not passing down the label so buildah a devpts which is not labeled correctly?

nalind commented:
As near as the `buildah` process can tell, SELinux is disabled, and `github.com/opencontainers/selinux/go-selinux/label.InitLabels()` returns empty values for it to use.

rhatdan commented:
Just for grins could you hard code a label into mountlabel, to see if the AVC's get better?

nalind commented:
When setting processLabel to `system_u:system_r:container_t:s0:c26,c85` and mountLabel to `system_u:object_r:container_file_t:s0:c26,c85`, runc's validator logic returns an `selinux label is specified in config, but selinux is disabled or not supported` error.

rhatdan commented:
Ok We will have to think of planb.  I had a patch for selinux to allow the applying of labels even if selinux is disabled, but maybe that will not work.

rhatdan commented:
@nalind Needs a rebase.

nalind committed:
Add and implement IsolationOCIRootless

Add an IsolationOCIRootless that runs the OCI runtime with its
--rootless flag, with network and UTS namespaces disabled, with IPC,
PID, and user namespaces forcibly enabled.  In this mode, we don't
attempt to set the container's hostname (because we don't have our own
UTS namespace), and we don't try to set any supplemental groups.  The
/sys directory is replaced with a bind mount of the host's /sys rather
than a fresh sysfs instance.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
.travis.yml: run integration tests with BUILDAH_ISOLATION=chroot

Run integration tests with both BUILDAH_ISOLATION=rootless, in addition
to previously-enabled tests for "oci" and "chroot".

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind commented:
bot, retest this please

rhatdan commented:
bot, retest this please

rhatdan commented:
bot, retest this please

nalind commented:
bot, retest this please

TomSweeneyRedHat commented on run.go:
```diff
@@ -1070,35 +1070,151 @@ func (b *Builder) Run(command []string, options RunOptions) error {
+	switch isolation {
+	case IsolationOCIRootless:
+		if ns := options.NamespaceOptions.Find(string(specs.IPCNamespace)); ns == nil || ns.Host {
+			logrus.Debugf("Forcing use of an IPC namespace.")
```
Just wondering if it would be helpful to have the value of the specs.IPCNamespace here, and the corresponding values in the debugs below.  I'm fine with this being a follow up PR.

nalind replied:
specs.IPCNamespace is the literal string "ipc".  Or do you mean the path of a previously-created IPC namespace, if one had been specified and was about to be ignored?

TomSweeneyRedHat replied:
My bad, I thought specs.IPCNamespace had a value of a created IPC namespace in it, rather than just "IPC".  I thought displaying the id/name of each of these values would be useful, but if they're not available without a lot of hoop jumping, ignore the comment.  Thanks for the 411.

TomSweeneyRedHat reviewed (commented):
Nit: I see MoreCreateArgs being set, but not referenced?  Is that future use or did I miss it?  The name could be a little better, ie. if it's going to track the socket location, ConsoleSockertArgs or some such.

TomSweeneyRedHat commented:
Happy green test buttons.  I'm fine with addressing my comments in a follow up if they're pertinent.

nalind commented:
The `MoreCreateArgs` value is passed to the child [here](https://github.com/projectatomic/buildah/pull/873/files#diff-12f3273da1625fbe0b8ae7f4eefe2c3cR1297), so that its values can be used as additional command line arguments for the runtime's `create` command.  Arguments that specify where to send the console socket are appended to the passed-in value instead of its previous empty default value.

TomSweeneyRedHat commented:
Yeah I saw the MoreCreateArgs getting passed, I just didn't find any code that actually used it once passed. I may have missed it or have older code.

rhatdan commented:
@rh-atomic-bot r+

rhatdan commented:
@rh-atomic-bot retry
